include/zephyr/spinlock.h

```diff
@@ -265,6 +265,20 @@ static ALWAYS_INLINE void k_spin_unlock(struct k_spinlock *l,
  * @cond INTERNAL_HIDDEN
  */
 
+/**
+ * @brief Checks if spinlock is held by some CPU, including the local CPU.
+ *		This API shouldn't be used outside the tests for spinlock
+ *
+ * @param l A pointer to the spinlock
+ * @retval true - if spinlock is held by some CPU; false - otherwise
+ */
+#if (defined(CONFIG_SMP) && defined(CONFIG_TEST)) || defined(__DOXYGEN__)
```

Comment:
Hi @andyross,

> Not sure why TEST is being tested here? If it's an API it should be an API for everyone. Also if you put the #if's inside the function and not outside, you don't need to special-case doc generation (it just turns into an empty inline for everyone else).

We introduce `k_spin_is_locked` only for the testing purposes - so we can check spinlock state in tests without usage of spinlock structure internal fields. This is usefull abstration as in https://github.com/zephyrproject-rtos/zephyr/pull/61541 we adding another spinlock implementation with different internal fields.

However I don't see any valid use-case of this API in real application - there always be a race conditions - even if we try to use it with interrupts disabled (as other core still may lock / unlock this spinlock).

That's why we have `CONFIG_TEST` guard - to avoid using this API in real applications.